python/ray/train/lightning/lightning_trainer.py

```diff
@@ -0,0 +1,374 @@
+    val_dataloaders = trainer_fit_params.get("val_dataloaders", None)
+
+    train_ray_dataset = session.get_dataset_shard("train")
+    val_ray_dataset = session.get_dataset_shard("val")
+    if train_ray_dataset:
+        if datamodule:
+            logger.warning(
+                "Using Ray datasets as primary input. The 'datamodule' defined in "
+                "'LightningConfig.trainer_fit_params' will be ignored!"
+            )
+
+        datamodule = RayDataModule(
+            dataset_iter_config=datasets_iter_config,
+            train_dataset=train_ray_dataset,
+            val_dataset=val_ray_dataset,
+        )
+
+    # Prepare Lightning Module
+    lightning_module = module_class(**module_init_config)
+
+    # Prepare Lightning Trainer
+    trainer_config["enable_progress_bar"] = False
+
+    # Setup trainer's parallel devices
+    if trainer_config.get("accelerator", None) == "gpu":
+        current_device = ray.train.torch.get_device()
+        trainer_config["devices"] = [current_device.index]
+
+    # Setup ray cluster environment info
+    trainer_config["plugins"] = [
+        plugin
+        for plugin in trainer_config.get("plugins", [])
+        if not isinstance(plugin, ClusterEnvironment)
+    ]
+    trainer_config["plugins"].append(RayEnvironment())
+
+    # Setup ddp strategy for ray orchestration
+    if "strategy" in trainer_config:
+        logger.warning(
+            "`strategy` specified in `LightningConfig.trainer_init_config` "
+            "will be ignored. LightningTrainer will create a RayDDPStrategy "
+            "object based on `LightningConfig.ddp_strategy_config`."
+        )
+    trainer_config["strategy"] = RayDDPStrategy(**ddp_strategy_config)
+
+    # TODO(yunxuanx): Next PR, add logging and checkpointing support
+    trainer_config["enable_checkpointing"] = False
+
+    # Filter out existing ModelCheckpoint Callbacks
+    callbacks = []
+    for callback in trainer_config.get("callbacks", []):
+        if isinstance(callback, ModelCheckpoint):
+            logger.warning(
+                "LightningTrainer will create a Ray ModelCheckpoint callback "
+                "based on `LightningConfig.model_checkpoint_config`. All the other"
+                " ModelCheckpoint callbacks are ignored."
+            )
+        else:
+            callbacks.append(callback)
+    trainer_config["callbacks"] = callbacks
+
+    trainer = pl.Trainer(**trainer_config)
+
+    # Restore the training from a previously interrupted/failed run.
+    ckpt_path = None
+    checkpoint = session.get_checkpoint()
+    if checkpoint:
+        ckpt_dir = checkpoint.to_directory()
+        ckpt_path = f"{ckpt_dir}/{MODEL_KEY}"
+
+    trainer.fit(
+        lightning_module,
+        datamodule=datamodule,
+        train_dataloaders=train_dataloaders,
+        val_dataloaders=val_dataloaders,
+        ckpt_path=ckpt_path,
+    )
+
+    if ckpt_path:
+        shutil.rmtree(os.path.dirname(ckpt_path))
```

Comment:
Instead of manually deleting, I think we can wrap the entire `trainer.fit` in a `with checkpoint.as_directory()`, which should have the same functionality I believe.

Reply:
I really love this idea, the code should be cleaner. We may need another trainer.fit() if there's no checkpoint to restore.

```python
if checkpoint:
    with checkpoint.as_directory() as ckpt_dir:
        trainer.fit()
else:
    trainer.fit()
```

Reply:
actually, why is the ``ckpt_dir`` not used?

Reply:
```
if checkpoint:
    with checkpoint.as_directory() as ckpt_dir:
        trainer.fit(model, ..., ckpt_path=f"{ckpt_dir}/{MODEL_KEY}")
else:
    trainer.fit(model, ..., )
```
Sorry for the confusion, I ignored the parameter list.